kaggle_environments/envs/halite/halite.js

```diff
@@ -455,21 +495,84 @@ async function renderer({
+
+  const getHalite = player => player[0];
+  const getCargo = player => Object.entries(player[2]).map(([, v]) => v[1]).reduce((a, b) => a + b, 0);
+  const getNumShips = player => Object.entries(player[2]).length;
+  const getNumShipyards = player => Object.entries(player[1]).length;
+  
+  // Writes two lines, "Halite" and "Cargo", and returns y value for what would be the third line.
+  const writeScoreboardText = (ctx, player, x, y) => {
+    ctx.fillText(`Halite: ${getHalite(player)}`, x, y);
+    ctx.fillText(`Cargo: ${getCargo(player)}`, x, y + scoreboardLineYDiffPx);
+    return y + 2 * scoreboardLineYDiffPx;
+  }
+
+  const scoreboardShipSizePx = scoreboardFontSizePx * 1.7;
+  const drawShip = (ctx, playerIndex, x, y, iconSize = scoreboardShipSizePx) => ctx.drawImage(
+    bufferCanvas, 500 + 100 * playerIndex, 0, fixedCellSize, fixedCellSize,
+    x, y, iconSize, iconSize);
+  const drawShipYard = (ctx, playerIndex, x, y, iconSize = scoreboardShipSizePx) => ctx.drawImage(
+    bufferCanvas, 500 + 100 * playerIndex, 400, fixedCellSize, fixedCellSize,
+    x, y, iconSize, iconSize);
+
+  const scoreboardShipXPaddingPx = scoreboardShipSizePx + scoreboardPaddingPx;
+  const drawShipAndYardCounts = (ctx, player, playerIndex, x, y, iconSize = scoreboardShipSizePx) => {
+    drawShip(ctx, playerIndex, x, y);
+    ctx.fillText(`x ${getNumShips(player)}`, x + scoreboardShipXPaddingPx, y + 0.28 * iconSize);
+    drawShipYard(ctx, playerIndex, x, y + iconSize);
+    ctx.fillText(`x ${getNumShipyards(player)}`, x + scoreboardShipXPaddingPx, y + 1.38 * iconSize);
+  }
+
+  // Render Scoreboard for each player, if we have enough room on the sides of the window.
+  if (width / height >= 1.3) {
+    fgCtx.fillStyle = "#FFFFFF";
+    fgCtx.font = `normal ${scoreboardFontSizePx}px sans-serif`;
+    fgCtx.textBaseline = "top";
+    fgCtx.textAlign = "left";
+    const topStartY = topLeftCell.dy;
+    const bottomStartY = botRightCell.dy + botRightCell.ds - 2 * scoreboardShipSizePx - 2 * scoreboardLineYDiffPx;
+    players.forEach((player, playerIndex) => {
+      const x = playerIndex % 2 === 1
+        ? Math.max(
+            // Make sure we don't start within the game area on the right side.
+            botRightCell.dx + botRightCell.ds + 2 * boxPadding,
+            width - topLeftCell.dy - 5.5 * scoreboardFontSizePx)
+        : topLeftCell.dy;
+      const startY = playerIndex < 2 ? topStartY : bottomStartY;
+      const nextY = writeScoreboardText(fgCtx, player, x, startY);
+      drawShipAndYardCounts(fgCtx, player, playerIndex, x, nextY);
     });
-  });
+  }
+
+  // Populate the legend which renders agent icons and names (see player.html).
+  if (agents && agents.length && (!agents[0].color || !agents[0].image)) {
+    const getPieceImage = playerIndex => {
+      const pieceCanvas = document.createElement("canvas");
+      parent.appendChild(pieceCanvas);
+      pieceCanvas.style.marginLeft = "10000px";
```

Comment:
what does this large left margin do?

Reply:
I copied this from the `connectx` legend logic, but I presume it's just making sure it's wayyyy off-screen on the parent element where I guess the raw image "lives" before copied into the Legend area.

Reply:
It's a bit confusing because the element is removed from the parent below in this function, so I'm not sure if the shift is really necessary, but I'm deferring to past-Adam.